### PR TITLE
feat(scaler): report status per scaler

### LINKED ISSUE #256: [FEAT] List application status messages in a list
At the moment, all the status messages are listed in a big ol' string. This makes scanning the information difficult both in the cli and in other interfaces that consume this data (i.e. wash ui).

I'd like to see the addition of a `status_message_array` property on the application status either as an array of strings, or an array of hashmaps with some extra info like time and priority.

### Existing

```json
{
  "apps": [
    {
      "deployed_version": "v0.0.1",
      "description": "...",
      "name": "kvcounter-rust",
      "status": "deployed",
      "status_message": "Could not satisfy spread default for file:///Users/lheywood/Projects/wasmCloud/wasmCloud/examples/rust/actors/http-hello-world/build/http_hello_world_s.wasm, 0/1 eligible hosts found., Putting link definition between kvcounter_rust-file____users_lheywood_projects_wasmcloud_wasmcloud_examples_rust_actors_http_hello_world_build_http_hello_world_s_wasm and kvcounter_rust-ghcr_io_wasmcloud_keyvalue_redis_0_23_0, Putting link definition between kvcounter_rust-ghcr_io_wasmcloud_http_server_0_20_0 and kvcounter_rust-file____users_lheywood_projects_wasmcloud_wasmcloud_examples_rust_actors_http_hello_world_build_http_hello_world_s_wasm, Could not satisfy spread default for ghcr.io/wasmcloud/http-server:0.20.0, 0/1 eligible hosts found., Could not satisfy spread default for ghcr.io/wasmcloud/keyvalue-redis:0.23.0, 0/1 eligible hosts found.",
      "version": "v0.0.1"
    }
  ],
  "success": true
}
```

### Proposed

```json
{
  "apps": [
    {
      "deployed_version": "v0.0.1",
      "description": "...",
      "name": "kvcounter-rust",
      "status": "deployed",
      "status_message": "Could not satisfy spread default for file:///Users/lheywood/Projects/wasmCloud/wasmCloud/examples/rust/actors/http-hello-world/build/http_hello_world_s.wasm, 0/1 eligible hosts found., Putting link definition between kvcounter_rust-file____users_lheywood_projects_wasmcloud_wasmcloud_examples_rust_actors_http_hello_world_build_http_hello_world_s_wasm and kvcounter_rust-ghcr_io_wasmcloud_keyvalue_redis_0_23_0, Putting link definition between kvcounter_rust-ghcr_io_wasmcloud_http_server_0_20_0 and kvcounter_rust-file____users_lheywood_projects_wasmcloud_wasmcloud_examples_rust_actors_http_hello_world_build_http_hello_world_s_wasm, Could not satisfy spread default for ghcr.io/wasmcloud/http-server:0.20.0, 0/1 eligible hosts found., Could not satisfy spread default for ghcr.io/wasmcloud/keyvalue-redis:0.23.0, 0/1 eligible hosts found.",
      "status_message_array": [
        {
          "message": "Could not satisfy spread default for file:///Users/lheywood/Projects/wasmCloud/wasmCloud/examples/rust/actors/http-hello-world/build/http_hello_world_s.wasm, 0/1 eligible hosts found.",
          "time": "2021-09-01T12:00:00Z",
          "level": "critical"
        },
        {
          "message": "Putting link definition between kvcounter_rust-file____users_lheywood_projects_wasmcloud_wasmcloud_examples_rust_actors_http_hello_world_build_http_hello_world_s_wasm and kvcounter_rust-ghcr_io_wasmcloud_keyvalue_redis_0_23_0",
          "time": "2021-09-01T12:00:00Z",
          "level": "info"
        },
        {
          "message": "Putting link definition between kvcounter_rust-ghcr_io_wasmcloud_http_server_0_20_0 and kvcounter_rust-file____users_lheywood_projects_wasmcloud_wasmcloud_examples_rust_actors_http_hello_world_build_http_hello_world_s_wasm",
          "time": "2021-09-01T12:00:00Z",
          "level": "info"
        },
        {
          "message": "Could not satisfy spread default for ghcr.io/wasmcloud/http-server:0.20.0, 0/1 eligible hosts found.",
          "time": "2021-09-01T12:00:00Z",
          "level": "critical"
        },
        {
          "message": "Could not satisfy spread default for ghcr.io/wasmcloud/keyvalue-redis:0.23.0, 0/1 eligible hosts found.",
          "time": "2021-09-01T12:00:00Z",
          "level": "critical"
        }
      ]
      "version": "v0.0.1"
    }
  ],
  "success": true
}
```

### PR DESCRIPTION
**Pull request by brooksmtownsend**
## Feature or Problem
This PR changes the status that is published, retrieved, and returned on API operations from the `StatusInfo` struct to the `Status` struct, which now includes a detailed status of all of the scalers running for a particular manifest.

For example, our HTTP hello world wadm now reports a status like this:
```
{
  "version": "v0.0.1",
  "status": {
    "type": "reconciling",
    "message": "Putting link definition between rust_hello_world-httpserver and rust_hello_world-http_component"
  },
  "components": [],
  "scalers": [
    {
      "name": "Spread: rust_hello_world-http_component",
      "status": {
        "type": "deployed"
      }
    },
    {
      "name": "Link: rust_hello_world-httpserver -(wasi:http)-> rust_hello_world-http_component",
      "status": {
        "type": "reconciling",
        "message": "Putting link definition between rust_hello_world-httpserver and rust_hello_world-http_component"
      }
    },
    {
      "name": "Spread: rust_hello_world-httpserver",
      "status": {
        "type": "deployed"
      }
    }
  ]
}
```

When querying all applications, you'll see the high level status info as you did before. However, when querying the application status itself with the `wash app status` command, we should be able to improve the output from this:
```
➜ wash app status rust-hello-world

  Name               Deployed Version   Deploy Status   Status Message
  rust-hello-world   v0.0.1            Reconciling      "Event modified scaler state, reconciling"
```

to this (forthcoming PR):

```
➜ cargo run -- app status rust-hello-world 
rust-hello-world@v0.0.1 - Failed                                                                          
                                                                                                                                               
  Name                                          Kind           Status           
  http_component                                SpreadScaler   Deployed         
  http_component_sad                            SpreadScaler   Failed (*)       
  totally_unrelated_component                   SpreadScaler   Failed (*)       
  httpserver -(wasi:http)-> http_component      LinkScaler     Deployed         
  httpserver                                    SpreadScaler   Reconciling (*)  
                                                                                
  Status Messages                                                               
                                                                                
  http_component_sad
    └ Could not satisfy spread eastcoast for file:///Users/brooks/github.com/wasmcloud/wasmcloud/examples/rust/components/http-hello-world/build/http_hello_world_s.wasm, 0/1 eligible hosts found. Could not satisfy spread westcoast for file:///Users/brooks/github.com/wasmcloud/wasmcloud/examples/rust/components/http-hello-world/build/http_hello_world_s.wasm, 0/1 eligible hosts found.

  totally_unrelated_component
    └ Could not satisfy spread foobar for file:///Users/brooks/github.com/wasmcloud/wasmcloud/examples/rust/components/http-hello-world/build/http_hello_world_s.wasm, 0/1 eligible hosts found.

  httpserver
    └ Scaling provider on 1 host(s)
```

## Related Issues
Fixes #256
Lays the ground work for #225
Works on #346 

## Release Information
wadm 0.14.0
wadm-client 0.3.0
wadm-types 0.3.0

## Consumer Impact
<!---
Indicate the impact, if any, this change will have on other consumers, dependencies, or dependents. In other words, the "blast radius" of the impact of this change and what steps related projects may need to take in response to this.
--->

## Testing
<!---
Declare the testing information for this pull request
--->

### Unit Test(s)
<!---
Indicate if unit tests were added or modified, and if so, which ones 
--->

### Acceptance or Integration
<!---
Indicate any changes or additions to the acceptance or integration test suite 
--->

### Manual Verification
<!---
Mandatory. Indicate the steps that you took to verify that this pull request works 
--->

### CODE REVIEW

File: crates/wadm/src/server/handlers.rs [hunk 1]
```diff
@@ -248,13 +248,20 @@ impl<P: Publisher> Handler<P> {
         };
 
         for model in &mut data {
+            // TODO: we want to disregard the status if the lattice isn't being observed yet... how?
+            // Need some form of query to check if the lattice is being observed
+            #[allow(deprecated)]
```
Comment (brooksmtownsend) [inline crates/wadm/src/server/handlers.rs:253]: This TODO is why #346 isn't solved here, we need an indicator to know that a lattice isn't observed yet in order to _truly_ decide whether to take the last status or put in waiting. Was hoping to noodle with @thomastaylor312 on this one, but it doesn't need to be fixed in this PR.

File: crates/wadm/src/consumers/manager.rs [hunk 2]
```diff
@@ -83,6 +83,11 @@ pub trait WorkerCreator {
     ) -> anyhow::Result<Self::Output>;
 }
 
+#[async_trait::async_trait]
+pub trait ConcernedThing {
```
Comment (thomastaylor312) [inline crates/wadm/src/consumers/manager.rs:87]: Lol this name. Maybe change it to something like `Interest` or `ConcernedWith`

File: crates/wadm/src/scaler/configscaler.rs [hunk 3]
```diff
@@ -36,6 +36,14 @@ impl<C: ConfigSource + Send + Sync + Clone> Scaler for ConfigScaler<C> {
         &self.id
     }
 
+    fn kind(&self) -> &str {
+        "ConfigScaler"
```
Comment (thomastaylor312) [inline crates/wadm/src/scaler/configscaler.rs:40]: Should this be a const too?

File: crates/wadm/src/scaler/spreadscaler/link.rs [hunk 4]
```diff
@@ -62,6 +62,20 @@ where
         &self.id
     }
 
+    fn kind(&self) -> &str {
+        LINK_SCALER_TYPE
+    }
+
+    fn name(&self) -> String {
+        format!(
+            "{} -({}:{})-> {}",
```
Comment (thomastaylor312) [inline crates/wadm/src/scaler/spreadscaler/link.rs:71]: Oooh I like this

File: crates/wadm/src/server/storage.rs [hunk 5]
```diff
@@ -116,15 +118,21 @@ impl ModelStorage {
...
                     Some(Ok(ModelSummary {
                         name: model_name,
                         version: manifest.current_version().to_owned(),
                         description: manifest.get_current().description().map(|s| s.to_owned()),
                         deployed_version: manifest.get_deployed().map(|m| m.version().to_owned()),
-                        // TODO(thomastaylor312): Actually fetch the status info from the stored
-                        // manifest once we figure it out
-                        status: StatusType::default(),
-                        status_message: None,
+                        // Since we are just fetching the manifest from the store, we need to wait
```
Comment (thomastaylor312) [inline crates/wadm/src/server/storage.rs:127]: Should we define a different type for a list response? Adding more of these fields feels like it makes it even more misleading as we don't know what the current status is at all. These could all be deployed, but it returns "waiting." I think a client method can fetch additional status and possibly call something like `ModelSummary::from(list_item)`

**Review by thomastaylor312 — CHANGES_REQUESTED** [review]
One item of discussion and a naming request. Looking forward to this landing because it is real nice

File: crates/wadm/src/consumers/manager.rs (hunk 2 above)
Comment (brooksmtownsend) [reply crates/wadm/src/consumers/manager.rs:87]: Oh boy. This is leftover from some experimentation

File: crates/wadm/src/scaler/configscaler.rs (hunk 3 above)
Comment (brooksmtownsend) [reply crates/wadm/src/scaler/configscaler.rs:40]: Yeah const in the file works, I tried to make this a const in the trait but with our backoff wrapping we can't proxy it down to the scalers

File: crates/wadm/src/server/storage.rs (hunk 5 above)
Comment (brooksmtownsend) [reply crates/wadm/src/server/storage.rs:127]: Yeah, probably. I think we assumed that we would eventually infer this information from the stored manifest, but we just don't.

**Review by thomastaylor312 — APPROVED** [review]

**Pull request merged**